feat(sncast): generate default name when using account import

khayss opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Closes #2581 

## Introduced changes


<!-- A brief description of the changes -->
All changes made are focused on `sncast`.
In `crates/sncast/src/starknet_commands/account/import.rs`, I added a utility function to generate a random account name which is used to supply the account construct with a name where the user does not specify any to the  Also updated the import function in this file by changing the function parameter to accept an `Option<String>` type for `account` parameter. This way, users can import accounts without requiring `name` to be specified. Also, updated the `Import` struct `name` key to have a type of `Option<String>`.

In `crates/sncast/src/main.rs`, I updated the argument feed into the `import` function `account` parameter.

I updated the necessary documentation and logs.

-

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,6 +204,15 @@ fn get_private_key_from_input() -> Result<Felt> {
     parse_input_to_felt(&input)
 }
 
+fn generate_default_account_name() -> String {
+    let start = SystemTime::now();
+    let since_epoch = start
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards");
+    let timestamp = since_epoch.as_millis();
+    format!("account-{}", timestamp)
+}
+
```
Let's refactor this function to create account with sequential names (`account-1`, `account-2`, ...). Name with timestamp (e.g. `account-1729865041172`) is too complicated.
Our logic should find next available numer (and handle gaps, e.g. having `account-1` and `account-3` in file, imported account should receive name `account-2`).

You can see how accounts are read from the file [here](https://github.com/foundry-rs/starknet-foundry/blob/e5918db054d7264fd08e79dee1c33d6391009eda/crates/sncast/src/starknet_commands/account/mod.rs#L102).

khayss replied:
Alright. This was my initial plan. I'll fall back to it.

franciszekjob reviewed (changes requested):
1. Please fix linting.
2. Add test for importing account without `--name` flag.
3. Mention the change in `CHANGELOG.md`.

khayss commented:
Hello @franciszekjob 
In this draft PR, I have implemented a way to generate accounts in a more sequential and friendly manner as requested.
However, I feel there are some challenges and possibly bugs.

I implemented the feature by tracking the number of default accounts and incrementing or decrementing as new accounts are added or removed. Also, I keep track of deleted ids in an array this way, we can reuse ids and cover gaps.

This implementation meant the structure of how the account is stored had to be restructured. This affect features asides from the `import` as they also read and write to the same account.

I still believe, the best approach, is defining a custom type that all features use, this way, if any change is made to one area, it'll cause an error that can be caught at compile time.

Also, due to the use of `serde_json::Value` type which is quite loose, we can run into some bugs and break other features by updating one.

For example, the following code can cause unexpected bugs.
https://github.com/foundry-rs/starknet-foundry/blob/f7cfe3e7b6792acb824ba344d4c63f9853a58fc6/crates/sncast/src/starknet_commands/account/deploy.rs#L340

If for instance there's a network with the name `deployment`, the line of code below will assign different data to the same key as above, which means one will overwrite the other.
https://github.com/foundry-rs/starknet-foundry/blob/f7cfe3e7b6792acb824ba344d4c63f9853a58fc6/crates/sncast/src/starknet_commands/account/mod.rs#L114C4-L114C50

This shows the need to use a common source of truth so we're certain, there's no conflict.

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -80,6 +80,12 @@ pub async fn import(
     };
     let private_key = &SigningKey::from_secret_scalar(*private_key);
 
+    // let account = if let Some(name) = account {
+    //     name
+    // } else {
+    //     generate_default_account_name()
+    // };
+
```
We shouldn't keep comments unless they are related with todo or there is a good reason to keep it in the codebase.

franciszekjob commented on crates/sncast/src/starknet_commands/account/mod.rs:
```diff
@@ -21,6 +21,21 @@ pub mod deploy;
 pub mod import;
 pub mod list;
 
+// Account Structure
+// {
+//  accounts: {
+//      network_name: {
+//          account_name: {
+//              account details
+//              },
+//          }
+//      },
+//  settings: {
+//    default_account_count: u32,
+//    available_default_account_ids: Vec<u32>
+//      }
+//},
+
```
Same as above.

franciszekjob reviewed (changes requested):
First of all, please address the changes I requested in previous review (formatting, tests etc.)

The only thing this feature should include, is to generate a default name if user doesn't pass one. This should be achieved by one function, there's no need to introduce new account structures etc.

Also, this feature shouldn't affect account deletion and any other functionalities.
Please keep it simple :).

khayss commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -252,4 +305,18 @@ mod tests {
     }
+
+    #[test]
+    fn test_default_account_name_import() {
+        let accounts_file = Utf8PathBuf::from("tests/data/accounts/accounts.json");
+
+        let (account, account_id) = super::get_import_name(None, &accounts_file);
+        let (account_2, account_id_2) = super::get_import_name(None, &accounts_file);
+
+        assert_eq!(account, "account-1");
+        assert_eq!(account_id, Some(1));
+
+        assert_eq!(account_2, "account-2");
+        assert_eq!(account_id_2, Some(2));
+    }
```
For some reason, this test and all other tests in the module are not captured.

franciszekjob replied:
Please base on https://github.com/foundry-rs/starknet-foundry/blob/master/crates/sncast/tests/e2e/account/import.rs while adding tests.

khayss commented:
I have made the changes you asked for.

I believe the linting is correct.
Also updated the change-log

khayss commented:
Back to this @franciszekjob 
I see 2 checks are failing, one is a typo which I'll fix in a new commit very soon and the other is Cast test. This again goes to show how tricky it is to implement a default name that needs to be deterministic in a way with the current setup. The reason is, to achieve this, you need to somehow add a new structure (even the easiest one such as incremental ids without accounting for gaps) to the accounts file, and adding anything there disrupts the initial setup.

I'll be expecting your review to know next steps to take.
Thank you.

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,9 +201,57 @@ fn get_private_key_from_input() -> Result<Felt> {
+pub fn get_import_name(
+    import_name: Option<String>,
+    accounts_file: &Utf8PathBuf,
+) -> (String, Option<u32>) {
```
1. I think `generate_account_name` could be better.
2. The function should return only one value which is the account name.

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,9 +201,57 @@ fn get_private_key_from_input() -> Result<Felt> {
+        if !accounts_file.exists() {
+            std::fs::create_dir_all(accounts_file.clone().parent().unwrap())
+                .expect("Failed to create directory");
+            std::fs::write(accounts_file.clone(), "{}").expect("Failed to write to file");
+        }
+
+        let contents = std::fs::read_to_string(accounts_file.clone()).expect("Failed to read file");
+        let mut items: serde_json::Value = serde_json::from_str(&contents)
+            .map_err(|_| anyhow!("Failed to parse accounts file at = {}", accounts_file))
+            .unwrap();
+
+        let mut available_default_ids: Vec<u32> = serde_json::from_value(
+            items["default_account_settings"]["available_default_ids"].clone(),
+        )
+        .unwrap_or(vec![]);
+        let total_defaults: u32 =
+            serde_json::from_value(items["default_account_settings"]["total_defaults"].clone())
+                .unwrap_or(0);
+
+        let total_defaults = total_defaults + 1;
+        let default_id = if available_default_ids.is_empty() {
+            total_defaults
+        } else {
+            let min_id = *available_default_ids.iter().min().unwrap();
+            available_default_ids.retain(|x| *x != min_id);
+            min_id
+        };
+        items["default_account_settings"]["available_default_ids"] = json!(available_default_ids);
+        items["default_account_settings"]["total_defaults"] = json!(total_defaults);
+
+        std::fs::write(
+            accounts_file.clone(),
+            serde_json::to_string_pretty(&items).unwrap(),
+        )
+        .expect("Failed to write to file");
+
+        (format!("account-{default_id}"), Some(default_id))
+    }
+}
+
```
The idea of calculating newly added account index should be simplified, there's no need to store information about total ids etc. 

The idea could look as follows:
- read all accounts with name matching `account-{i}` pattern
- extract their numbers
- find the first free smallest number which will be then used for new account name

Is that idea clear to you?

franciszekjob commented:
Hi @khayss, please have a look and address the comments I've added. Thanks.

franciszekjob commented:
Hey @khayss , just wanted to check out how's the progress :) ?

khayss commented:
Hey @franciszekjob I'll make a PR on this soon. Got all the requests and suggestions.

khayss commented:
Hello, @franciszekjob Made the requested changes. I'll be awaiting your review :)

khayss commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -1,3 +1,5 @@
+use std::collections::HashSet;
```
I am using HashSet because it's faster considering there's the possibility of gaps between names. e.g. `account-1`, `account-2`, `account-3`. This way, we don't need to first sort if we were to use an array (`Vec`)

khayss commented on crates/sncast/src/starknet_commands/account/mod.rs:
```diff
@@ -21,6 +21,8 @@ pub mod deploy;
 pub mod import;
 pub mod list;
 
+type NestedMap<T> = HashMap<String, HashMap<String, T>>;
```
I moved this up here since it's also used in the `read_and_flatten` function `list.rs` and `generate_account_name` function in `import.rs`. Thus, preventing duplicity.

khayss commented:
I see some tests are failing. Even the previous ones. Something about not able to read `account.json` file. Any idea about this?

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,6 +201,30 @@ fn get_private_key_from_input() -> Result<Felt> {
     parse_input_to_felt(&input)
 }
 
+fn generate_account_name(accounts_file: &Utf8PathBuf) -> anyhow::Result<String> {
+    let networks: NestedMap<AccountData> = read_and_parse_json_file(accounts_file)?;
+    let mut result = HashSet::new();
+
+    for (_network, accounts) in networks {
+        for (name, _data) in accounts {
+            if let Some(id) = name
+                .strip_prefix("account-")
+                .and_then(|id| id.parse::<u32>().ok())
+            {
+                result.insert(id);
+            };
+        }
+    }
+
+    let mut id = 1;
+
+    while result.contains(&id) {
+        id += 1;
+    }
+
+    Ok(format!("account-{id}"))
+}
+
```
The logic here is good, it just needs to handle the case when there is no json file with the accounts. In such case, `read_and_parse_json_file` will return `Err`. We should handle it and return `Ok(format!("account-{id}"))` (`id` should be equal to `1` at that time, because it will be first imported account, obvious).

It could look like this:
```rust
fn generate_account_name(accounts_file: &Utf8PathBuf) -> anyhow::Result<String> {
    let mut id: u32 = 1;
    let networks: NestedMap<AccountData> = match read_and_parse_json_file(accounts_file) {
        Ok(networks) => networks,
        Err(_) => return Ok(format!("account-{id}")),
    };
    
    let mut result = HashSet::new();

    for (_network, accounts) in networks {
        for (name, _data) in accounts {
            if let Some(id) = name
                .strip_prefix("account-")
                .and_then(|id| id.parse::<u32>().ok())
            {
                result.insert(id);
            };
        }
    }

    while result.contains(&id) {
        id += 1;
    }

    Ok(format!("account-{id}"))
}
```

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,58 @@ pub async fn test_invalid_address_computation() {
     "});
 }
+
+#[test_case("oz", "open_zeppelin"; "oz_account_type")]
+#[test_case("argent", "argent"; "argent_account_type")]
+#[test_case("braavos", "braavos"; "braavos_account_type")]
```
No need to put it here, we can hardcode type to `"oz"`.

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,58 @@ pub async fn test_invalid_address_computation() {
+        "--class-hash",
+        DEVNET_OZ_CLASS_HASH_CAIRO_0,
+        "--type",
+        input_account_type,
```
Please change this to `"oz"`.

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,58 @@ pub async fn test_invalid_address_computation() {
+                    "legacy": true,
+                    "private_key": "0x456",
+                    "public_key": "0x5f679dacd8278105bd3b84a15548fe84079068276b0e84d6cc093eb5430f063",
+                    "type": saved_type
```
Should be changed to `"open_zeppelin`".

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,58 @@ pub async fn test_invalid_address_computation() {
+#[test_case("argent", "argent"; "argent_account_type")]
+#[test_case("braavos", "braavos"; "braavos_account_type")]
+#[tokio::test]
+pub async fn test_happy_case_unnamed_import(input_account_type: &str, saved_type: &str) {
```
After above comment, these params should be removed as well.

franciszekjob reviewed (changes requested):
Looks good 👍 
Please address the changes I requested and also refactor test. 
The test should import e.g. 3 accounts, then remove `account-2` and then use import again. We should check if first three imported accounts have correct names and whether re-import will result in correct name (should be `account-2, to fill the gap).

khayss commented:
Made the requested changes and refactored the test.

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,97 @@ pub async fn test_invalid_address_computation() {
+    "});
+    }
+
+    let snapbox = runner(&delete_args).current_dir(tempdir.path());
+    snapbox.assert().success().stdout_matches(indoc! {r"
+        command: account delete
+        result: Account successfully removed
+    "});
+
```
Let's check the accounts file content here. It should only have `account-1` and `account-3`.

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,97 @@ pub async fn test_invalid_address_computation() {
 }
+
+#[tokio::test]
+pub async fn test_happy_case_unnamed_import() {
```
nit:
```suggestion
pub async fn test_happy_case_default_name_generation() {
```

khayss replied:
renamed. Thanks :)

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,97 @@ pub async fn test_invalid_address_computation() {
+                    "public_key": "0x5f679dacd8278105bd3b84a15548fe84079068276b0e84d6cc093eb5430f063",
+                    "type": "open_zeppelin"
+                  },
+                    "account-3": {
+                        "address": "0x123",
+                        "class_hash": DEVNET_OZ_CLASS_HASH_CAIRO_0,
+                        "deployed": false,
+                        "legacy": true,
+                        "private_key": "0x456",
+                        "public_key": "0x5f679dacd8278105bd3b84a15548fe84079068276b0e84d6cc093eb5430f063",
+                        "type": "open_zeppelin"
+                    }
```
Please fix indentation here.

khayss replied:
fixed

franciszekjob commented on crates/sncast/tests/e2e/account/import.rs:
```diff
@@ -683,3 +683,97 @@ pub async fn test_invalid_address_computation() {
+    let contents = fs::read_to_string(tempdir.path().join(accounts_file))
+        .expect("Unable to read created file");
+    let contents_json: serde_json::Value = serde_json::from_str(&contents).unwrap();
+    assert_eq!(
```
```suggestion
    assert_eq(
```

khayss replied:
I tried removing the exclamation mark and it threw an error, so I kept it.

franciszekjob reviewed (changes requested):
It's almost good to go 😄 .

Just please address the changes and fix test.

The test should work this way:
- import 3 accounts
- check the file contents
- delete `account-2`
- check the file contents
- import account again
- check the file contents

khayss commented:
reworked the test

khayss commented:
updated the docs.

franciszekjob commented on CHANGELOG.md:
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+#### Added
+
+- Feature to import an account without requiring the `--name` flag. This will under the hood generate a default name for the account This allows for easier and more streamlined account imports.
```
Actually I liked the previous description more. Could you please bring it back?

khayss commented:
I have made the updates @franciszekjob 
The tests and linting should pass now.
Also updated the change-log

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,6 +201,33 @@ fn get_private_key_from_input() -> Result<Felt> {
     parse_input_to_felt(&input)
 }
 
+fn generate_account_name(accounts_file: &Utf8PathBuf) -> String {
+    let mut id = 1;
+
+    let networks: NestedMap<AccountData> = match read_and_parse_json_file(accounts_file) {
+        Ok(networks) => networks,
+        Err(_) => return format!("account-{id}"),
+    };
+    let mut result = HashSet::new();
+
+    for (_network, accounts) in networks {
+        for (name, _data) in accounts {
+            if let Some(id) = name
+                .strip_prefix("account-")
+                .and_then(|id| id.parse::<u32>().ok())
+            {
+                result.insert(id);
+            };
+        }
+    }
+
+    while result.contains(&id) {
+        id += 1;
+    }
+
+    format!("account-{id}")
+}
```
```suggestion
fn generate_account_name(accounts_file: &Utf8PathBuf) -> Result<String> {
    let mut id = 1;

    if !accounts_file.exists() {
        return Ok(format!("account-{id}"));
    }

    let networks: NestedMap<AccountData> = read_and_parse_json_file(accounts_file)?;
    let mut result = HashSet::new();

    for (_network, accounts) in networks {
        for (name, _data) in accounts {
            if let Some(id) = name
                .strip_prefix("account-")
                .and_then(|id| id.parse::<u32>().ok())
            {
                result.insert(id);
            };
        }
    }

    while result.contains(&id) {
        id += 1;
    }

    Ok(format!("account-{id}"))
}
```

franciszekjob commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -80,6 +83,8 @@ pub async fn import(
     };
     let private_key = &SigningKey::from_secret_scalar(*private_key);
 
+    let account = account.unwrap_or(generate_account_name(accounts_file));
```
```suggestion
    let account = account.unwrap_or(generate_account_name(accounts_file)?);
```

franciszekjob reviewed (changes requested):
Looks good! Please address suggestions that I've left and we we'll be close to merge.  

@cptartur @kkawula @ddoktorski could someone of you have a look as well?

khayss commented:
Made the requested changes

cptartur commented on CHANGELOG.md:
```diff
@@ -11,6 +11,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Added
 
+- You can skip `--name` flag when using `account import` - a default name will be generated.
+
+## [Unreleased]
+
+### Cast
+
+#### Added
```
You've duplicated the unreleased section

khayss replied:
Oopsie. I will remove it.

khayss replied:
There's another unreleased for a different feature added there. Seems to be different and not a duplicate

cptartur replied:
There should be one `[Unreleased]` section. We change this into a relevant version when releasing.

khayss replied:
Got it. Thanks

cptartur commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -196,6 +201,34 @@ fn get_private_key_from_input() -> Result<Felt> {
+    let networks: NestedMap<AccountData> = read_and_parse_json_file(accounts_file)?;
+    let mut result = HashSet::new();
+
+    for (_network, accounts) in networks {
+        for (name, _data) in accounts {
```
```suggestion
    for (_, accounts) in networks {
        for (name, _) in accounts {
```

khayss replied:
I'll effect this suggestion.

cptartur commented on docs/src/starknet/account-import.md:
````diff
@@ -136,4 +136,17 @@ $ sncast \
     --type oz
+```
+
+#### Auto-generated Account Name
+
+To import an account with an auto-generated name, skip the `--name` flag.
+
+```shell
+$ sncast \
+    account import \
+    --url http://127.0.0.1:5050 \
+    --address 0x1 \
+    --private-key 0x2 \
+    --type oz
 ```
````
Instead of this, I'd, under the shell in `#### General Example`, add something like this "If the `--name` is not provided a default name will be generated instead.

cptartur reviewed (commented):
One problem is see with this is that how does the user know what name was generated for them? Does it show in the console?

khayss commented:
Yeah, I'll make the updates ASAP

khayss commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -142,30 +149,27 @@ pub async fn import(
-        } else {
-            "--add-profile flag was not set. No profile added to snfoundry.toml".to_string()
-        },
+        add_profile: import.add_profile.as_ref().map_or_else(
+            || "--add-profile flag was not set. No profile added to snfoundry.toml".to_string(),
+            |profile_name| format!("Profile {profile_name} successfully added to snfoundry.toml"),
+        ),
+        account_name: account.map_or_else(
+            || Some(format!("Account imported with name: {account_name}")),
+            |_| None,
+        ),
```
Due to the limitation on the acceptable lines of code for a function, I have to refactor this to save up some lines as `cargo clippy` will flag the `import` function as having too many lines (>100).

khayss commented on crates/sncast/src/response/structs.rs:
```diff
@@ -66,6 +66,7 @@ impl CommandResponse for AccountCreateResponse {}
 #[derive(Serialize)]
 pub struct AccountImportResponse {
     pub add_profile: String,
+    pub account_name: Option<String>,
```
This is added so we can show the generated account name on the terminal when `--name` flag is skipped.

khayss commented:
I have made the requested changes. Kindly review.

franciszekjob commented on CHANGELOG.md:
```diff
@@ -13,6 +13,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Added
 
+- You can skip `--name` flag when using `account import` - a default name will be generated.
+
+### Cast
+
+#### Added
+
```
```suggestion
- You can skip `--name` flag when using `account import` - a default name will be generated.
```

franciszekjob reviewed (changes requested):
Please apply changes.

khayss commented on crates/sncast/src/starknet_commands/account/import.rs:
```diff
@@ -80,6 +83,10 @@ pub async fn import(
 
+    let account_name = account
+        .clone()
+        .unwrap_or_else(|| generate_account_name(accounts_file).unwrap());
```
I'm using `unwrap_or_else` here because, `unwrap_or` is eagerly compiled and since we were using the return of a function call in the `unwrap_or`, it means, the function will be evaluated even when we don't need the value.

franciszekjob commented:
Lets wait for @cptartur approve as well.

cptartur commented on docs/src/starknet/account.md:
````diff
@@ -88,6 +88,7 @@ $ sncast \
     --private-key 0x2 \
     --type oz
 ```
+The `--name` can be omitted as this is optional. A default name will be generated for the account.
````
I'd not add that here, just in the other file you added.

khayss replied:
removed

cptartur commented on docs/src/starknet/account-import.md:
````diff
@@ -80,6 +80,7 @@ $ sncast \
     --private-key 0x2 \
     --type oz
 ```
+The `--name` can be omitted as this is optional. A default name will be generated for the account.
````
```suggestion
> 📝 **Note** 
> The `--name` can be omitted as this is optional. A default name will be generated for the account.
```

khayss replied:
Updated.

cptartur commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Added
 
+- You can skip `--name` flag when using `account import` - a default name will be generated.
```
Let's move this to `[Unreleased]` section

cptartur commented:
@khayss Please just address the suggested changes and we are good to merge.

khayss commented:
@cptartur Made the requested changes.

franciszekjob commented:
Thanks for contribution @khayss!

khayss commented:
My pleasure. Thanks for the great support throughout.
